daemon: Log Envoy version at startup.

jrajahalme opened the pull request:
Run the "cilium-envoy" binary with the "--version" option at the
startup and log version string that is output. Cause cilium-agent to
fail to start if "cilium-envoy" cannot be executed.

Fixes: #2632
Requested-by: André Martins <andre@cilium.io>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

aanm commented on pkg/envoy/envoy.go:
```diff
@@ -94,6 +94,21 @@ type Logger interface {
+func LogEnvoyVersion() {
+	out, err := exec.Command("cilium-envoy", "--version").Output()
+	if err != nil {
+		log.WithError(err).Fatal(`Envoy binary "cilium-envoy" cannot be executed`)
```
Should we be more strict and only allow a particular version to be executed? (I've a feeling this is not a good idea for developers but might be a good idea for releases?)

raybejjani replied:
Do we allow installations where the user provides envoy? If not then it might not matter? (I don't think we do, since we have our own module but we also call it cilium-envoy)

aanm replied:
I was just thinking where a developer might be running an old and incompatible version of cilium-envoy without realizing.

raybejjani commented on pkg/envoy/envoy.go:
```diff
@@ -94,6 +94,21 @@ type Logger interface {
+	for ; start < end && out[start] == '\n'; start++ {
+	}
+	for ; start < end && out[end-1] == '\n'; end-- {
+	}
```
Could you add comment explaining what we're trying to do here? It's not apparent that the version output has the leading and trailing newlines.
Alternatively, a regex to match the line, and extract the version directly, might be more resilient? In particular, if the version output ever added another line that would print oddly in the Infof below.

aanm replied:
Or https://golang.org/pkg/strings/#TrimSpace

jrajahalme replied:
Nice, using this now!

raybejjani commented on daemon/main.go:
```diff
@@ -500,6 +501,18 @@ func initEnv(cmd *cobra.Command) {
+	}
+	// Make sure Envoy version matches ours
+	if !strings.HasPrefix(envoyVersionArray[2], version.GetCiliumVersion().Revision) {
+		log.Fatal("Envoy version mismatch, aborting.")
```
If we make the mismatch issue just a warning we don't need a disableEnvoyVersionCheck flag thing just for tests (unless we allow both the prod version and whatever the tests provide). It would also meet aanm's requirement to allow a developer to run a modified version. Up to you, though.

aanm replied:
Yeah, I think this is the best option overall.

jrajahalme replied:
@tgraf @joestringer On a closer look this does not affect unit testing at all, and bash tests seem be fine in this regard as well. The test failure is due to a false positive potential deadlock due to kv-store transaction(s) taking a long time. The forthcoming kv-store improvements should allow us to remove the kv-store roundtrips we now do while holding a buildmutex.

To verify this I'll ask for re-test. If it passes then we could also keep the version mismatch as a fatal event, at least for now so that we catch any situations where it could happen?

jrajahalme committed:
daemon: Log and verify Envoy version at startup.

Run the "cilium-envoy" binary with the "--version" option at the
startup and log version string that is output. Cause cilium-agent to
fail to start if "cilium-envoy" cannot be executed or if Envoy version
does not match with the cilium-agent version.

To enable this a bug in Makefile that appended the git SHA to
envoy/SOURCE_VERSION is fixed.

Finally, envoy/Makefile now forces re-linking Envoy binary at each
invocation to ensure that the version SHA is fresh.

Requested-by: André Martins <andre@cilium.io>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
